iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java

```diff
@@ -282,6 +311,98 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
+      // get field name from original schema using source id of partition field
+      // because Pspec appends _bucket and _trunc to field name for bucket and truncate fields
+      String fieldName = originalSchema.findField(field.sourceId()).name();
+      // Check if the partition field is present in new schema
+      if (newSchema.findField(fieldName) == null) {
+        throw new IllegalArgumentException(
+            "Field " + fieldName + " does not exist in the new schema");
+      }
+
+      // Recreate the transform using the string representation
+      String transformString = field.transform().toString();
+
+      // Add the field to the new PartitionSpec based on the transform type
+      if ("identity".equalsIgnoreCase(transformString)) {
+        builder.identity(fieldName);
+      } else if (transformString.startsWith("bucket[")) {
+        // Extract bucket number from the string (e.g., bucket[16])
+        int numBuckets =
+            Integer.parseInt(
+                transformString.substring(
+                    transformString.indexOf('[') + 1, transformString.indexOf(']')));
+        builder.bucket(fieldName, numBuckets);
+      } else if (transformString.startsWith("truncate[")) {
+        // Extract width from the string (e.g., truncate[10])
+        int width =
+            Integer.parseInt(
+                transformString.substring(
+                    transformString.indexOf('[') + 1, transformString.indexOf(']')));
+        builder.truncate(fieldName, width);
+      } else if ("year".equalsIgnoreCase(transformString)) {
+        builder.year(fieldName);
+      } else if ("month".equalsIgnoreCase(transformString)) {
+        builder.month(fieldName);
+      } else if ("day".equalsIgnoreCase(transformString)) {
+        builder.day(fieldName);
+      } else if ("hour".equalsIgnoreCase(transformString)) {
+        builder.hour(fieldName);
+      } else {
+        throw new UnsupportedOperationException("Unsupported transform: " + transformString);
+      }
+    }
+
+    return builder.build();
+  }
+
+  /**
+   * Build a new sort order with new schema from original sort order. The new sort order has the
+   * same fields as the original sort order with source ids from the new schema
+   *
+   * @param originalSortOrder
+   * @param newSchema
+   * @return new SortOrder
+   */
+  public static SortOrder rebuildSortOrder(SortOrder originalSortOrder, Schema newSchema) {
+    SortOrder.Builder builder = SortOrder.builderFor(newSchema);
+
+    for (SortField field : originalSortOrder.fields()) {
+      // Find the field name in the original schema based on the sourceId
+      String fieldName = originalSortOrder.schema().findField(field.sourceId()).name();
+      // Check if the sortorder field is present in new schema
+      if (newSchema.findField(fieldName) == null) {
+        throw new IllegalArgumentException(
+            "Field " + fieldName + " does not exist in the new schema");
```

Comment:
what does this exception become when it reaches client ?

Reply:
IIUC, it get translated to a 400 error HttpStatus.BAD_RREQUEST

https://github.com/linkedin/openhouse/blob/main/services/common/src/main/java/com/linkedin/openhouse/common/exception/handler/OpenHouseExceptionHandler.java#L365

Reply:
sg.